x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java

```diff
@@ -652,149 +621,209 @@ private GetAliasesRequest createGetAliasesRequest(FieldCapabilitiesResponse resp
-            return emptyMap();
-        }
+    private static Map<String, InvalidMappedField> validateField(String fieldName, Map<String, QlFieldCapabilities> qlFieldCaps,
+        Map<String, FieldCapabilities> esTypes)
+    {
         Map<String, InvalidMappedField> invalidFields = new HashMap<>();
-        Map<String, Set<String>> typesErrors = new HashMap<>(); // map holding aliases and a list of unique field types across its indices
-        Map<String, Set<String>> aliasToIndices = new HashMap<>(); // map with aliases and their list of indices
 
-        Iterator<ObjectObjectCursor<String, List<AliasMetadata>>> iter = aliases.iterator();
-        while (iter.hasNext()) {
-            ObjectObjectCursor<String, List<AliasMetadata>> index = iter.next();
-            for (AliasMetadata aliasMetadata : index.value) {
-                String aliasName = aliasMetadata.alias();
-                aliasToIndices.putIfAbsent(aliasName, new HashSet<>());
-                aliasToIndices.get(aliasName).add(index.key);
+        // for each indices group
+        for (Entry<String, QlFieldCapabilities> capabilities : qlFieldCaps.entrySet()) {
+            String indicesGroupName = capabilities.getKey();
+            Map<String, QlFieldCapability> types = capabilities.getValue().getFields().get(fieldName);
+            
+            if (types == null || fieldName.startsWith("_")) {
+                continue;
+            }
+            InvalidMappedField invalidField = validate(fieldName, types);
+            if (invalidField != null) {
+                invalidFields.put(indicesGroupName, invalidField);
+            } else if (qlFieldCaps.size() == 1 && qlFieldCaps.containsKey("*") && esTypes.containsKey(KEYWORD.esType())
+                && esTypes.containsKey(CONSTANT_KEYWORD.esType())) {
+                esTypes.remove(CONSTANT_KEYWORD.esType());
             }
         }
 
-        // iterate over each type
-        for (Entry<String, FieldCapabilities> type : types.entrySet()) {
-            String esFieldType = type.getKey();
-            if (esFieldType == UNMAPPED) {
-                continue;
+        if (invalidFields.isEmpty() == false) {
+            return invalidFields;
+        }
+        // everything checks
+        return emptyMap();
+    }
+    
+    private static InvalidMappedField validate(String fieldName, Map<String, QlFieldCapability> types) {
+        StringBuilder errorMessage = new StringBuilder();
+        
+        boolean hasUnmapped = types.containsKey(UNMAPPED);
+        // a keyword field and a constant_keyword field with the same name in two different indices are considered "compatible"
+        // since a common use case of constant_keyword field involves two indices with a field having the same name: one being
+        // a keyword, the other being a constant_keyword
+        boolean hasCompatibleKeywords = types.containsKey(KEYWORD.esType()) && types.containsKey(CONSTANT_KEYWORD.esType());
```

Comment:
This check is already in `validateFields`.

Reply:
Hm... the other check is [here](https://github.com/elastic/elasticsearch/pull/55425/files#diff-e133dec5642088d77645639c836c858aR651) and there it's removing the "duplicate" field so that the final merged mapping has only one field with the same name and type `KEYWORD`.